### ISSUE TITLE
Use a coherent schema definition syntax

### ISSUE THREAD

**Issue by erdalkaraca** [bids-standard/bids-specification#884]
Currently, the schema consists of a mix of ad-hoc definitions and syntax.
This makes it hard to implement a tool to generically validate a given BIDS dataset.

In the ancpbids project (see below for link) we experimented with various schema definition languages, such as:
- XSD (XML Schema Definition): industry standard with mature support in many programming languages
- YAMALE: a community effort to formalize data structures using YAML format

Since the BIDS community is more comfortable with JSON/YAML formats, I would like to propose using a more formal schema definition language such as YAMALE to map the BIDS specification onto a machine readable schema.

Examples to follow as additional comments...

Links:
ancpbids: https://github.com/ANCPLabOldenburg/ancp-bids
XSD: https://www.w3.org/TR/xmlschema11-1/
Yamale: https://github.com/23andMe/Yamale

**Comment by erdalkaraca** [comment]
An example of how a coherent schema might look like using Yamale:

```yaml
Dataset:
  .extends: Folder
  subjects: list(include('Subject'), required=False)
  dataset_description: include('DatasetDescriptionFile')
  README: include('File', required=False)
  CHANGES: include('File', required=False)
  LICENSE: include('File', required=False)
  genetic_info: include('JsonFile', required=False)
  samples: include('JsonFile', required=False)
  participants_tsv: include('ParticipantsTsvFile', required=False)
  participants_json: include('TsvSidecarFile', required=False)
  code: include('Folder', required=False)
  derivatives: include('DerivativeFolder', required=False)
  sourcedata: include('Folder', required=False)
  stimuli: include('Folder', required=False)

---
JSONSchemaType:
  map()

MetadataFieldDefinition:
  name: str()
  description: str()
  type: include('JSONSchemaType')

SuffixDefinition:
  name: str()
  description: str()
  type: include('JSONSchemaType')

EntitiyDefinition:
  key: str()
  name: str()
  entity: str()
  description: str()
  type: include('JSONSchemaType')
```

**Comment by erdalkaraca** [comment]
Example definitions of entities:

```yaml
- key: subject
  name: Subject
  entity: sub
  description: |
    A person or animal participating in the study.
  type: label
- key: run
  name: Run
  entity: run
  description: |
    If several scans with [...]
  type: index
- key: mtransfer
  name: Magnetization Transfer
  entity: mt
  description: |
    If files belonging [...]
  type:
    enum:
      - "on"
      - "off"
- key: part
  name: Part
  entity: part
  description: |
    This entity is used to [...]
  type:
    enum:
      - mag
      - phase
      - real
      - imag
```

**Comment by erdalkaraca** [comment]
Example metadata fields definitiones:

```yaml
- name: AcquisitionDuration
  description: |
    Duration (in seconds) of [...]
  type:
    min: 0
    unit: s

- name: AnatomicalLandmarkCoordinates
  description: |
    Key:value pairs of any [...]
    example: `{"AC": [127,119,149], "PC": [128,93,141],
    "IH": [131,114,206]}`, or `{"NAS": [127,213,139], "LPA": [52,113,96],
    "RPA": [202,113,91]}`
  type:
    patternProperties:
      "^[A-Z]{2}$":
        type: array
          items:
            type: number
          minItems: 3
          maxItems: 3
```

**Comment by Tokazama** [comment]
This is something I've found a bit odd about BIDS. Personally, I would've gone with [JSON Schema](https://json-schema.org/) because it can actually be formalized into an already developed system for schemas. There may be some larger vision here that I don't understand though, so I'll go along with whatever working system is developed.

**Comment by erdalkaraca** [comment]
@Tokazama At https://github.com/ANCPLabOldenburg/ancp-bids with have experimented with several "systems for schemas":
- XSD, JSON Schema, YAMALE/Yaml

XSD is very verbose and "noisy" meaning it is not very convenient for humans to read just by using a text editor, but it is used in industries a lot and it has matured over decades.
XSD requires good (visual) tools support which are not freely available. Eclipse XML editors are open source but the installation/usage seems too complex for future contributors who just want to make simple modifications/proposals.

JSON is much more compact and used in machine-to-machine communication to transfer data objects (for example, web services, REST endpoints use it a lot), and it is still kind of readable for humans. Since JSON was meant for pure data transfer, it has no support for "decorating" specific lines, for example, it is lacking a built-in comment syntax, that is why nowadays you would be "hacking" in a synthetic property "_comment" which parsers/tools are to ignore when processing.

Yaml can be very compact but still has an exhaustive syntax.
The bad news about Yaml (at the moment) is that there is no standard "system for schemas" to use. To not come up with yet another schema definition format, we experimented with https://github.com/23andMe/Yamale which is what you see in the above comments.

The way the metadata fields are defined, we need a schema that allows to embed another schema. For example, see the above definition for AnatomicalLandmarkCoordinates where we have concrete data such as name and description, but we also need to describe (the structure of) the value object. It seems just fine to use a subset of JSON Schema for this.

In summary, there is no limiting factor to choose one schema language over the other as all can be transformed in a representation of each other, but from our experiments, the mix of Yamale and JSON Schema seems a good fit.

> This is something I've found a bit odd about BIDS. Personally, I would've gone with [JSON Schema](https://json-schema.org/) because it can actually be formalized into an already developed system for schemas. There may be some larger vision here that I don't understand though, so I'll go along with whatever working system is developed.

**Comment by Tokazama** [comment]
So is it correct to say that the biggest reason JSON isn't being used is because there's not a good way to add comments?

**Comment by erdalkaraca** [comment]
Well, as also mentioned, Yaml is more compact and has some more features that JSON does not provide such as mixing documents from different namespaces which is called TAGs in Yaml: https://yaml.org/spec/1.2/spec.html#id2782090
But not sure if there is a use case for that feature in the BIDS spec.

**Comment by Tokazama** [comment]
My understanding of TAGs in YAML (which I admit is very limited) is that they are most useful for telling specific applications parsing the file how to do something, which seems like it would be language specific and counterproductive to the end goal of a cross language schema. I'm not trying to say the current direction (using YAML) is bad. It's just that schemas have been around a while and we aren't looking for high performance I/O on these so I assume we just want a solid set of tools to build on rather than building up something new.

If you already have it figured out and I'm just eating up valuable development time then feel free to ignore this. It may be that I'm just completely missing the point.

**Comment by tsalo** [comment]
Sorry for the delay in responding. I don't think we will be using any features that are specific to YAML _beyond_ the file type, since YAML files are (1) more readable and (2) allow comments, so converting to valid JSON files should be trivial. For that reason, we're still looking into using JSON schema, but having a YAML-based schema definition language would definitely be nice!

@erdalkaraca your YAMALE-based approach looks really good, although we'll need to adjust based on changes from #883 and other recent PRs. I'm still not sure how many of the rules that we need to implement are possible with YAMALE (or JSON schema), but, as you said, adopting something like XSD is just not feasible within the community.

For now, I like the idea of adopting either YAMALE or JSON schema for the object validation (files in `schema/objects/`) and holding off on the files in `schema/rules/`, which don't cover enough of the actual rules in the specification for us to commit to any tools yet. At least not until we've made more progress with #620.

@effigies @rwblair do you have any thoughts on the proposed schema syntax?

**Comment by erdalkaraca** [comment]
> Perhaps a way of making that a bit more objective is the cost to switch to JSON. I've assumed that in the long run conforming with JSON schemas that are already established would save time in the long run, but if there is no actual benefit then we're just making breaking changes for the sake of opinion.

+1, but the schema is rather for technical people to implement software tools. One idea we had at our lab was to have a graphical (and interactive) representation of the schema, so non-technical people could navigate through that graph instead of the textual BIDS spec. At that level, it does not even matter what format/language is used behind the scenes...

**Comment by yarikoptic** [comment]
The most recent effort for a language to describe schemas is https://linkml.io which is getting adopted by NWB, DANDI and exported to by OpenMINDS. Added to the listing above.

**Comment by Tokazama** [comment]
Linked JSON is a good way to manage mainly schema related stuff. JSON isn't going anywhere and each time a new linked data standard comes out a corollary interface is adopted into JSON. If we are talking about table data and large arrays there is a lot more to consider.

**Comment by yarikoptic** [comment]
What do you mean exactly by "Linked JSON" @Tokazama? json-ld?

**Comment by Tokazama** [comment]
Most JSON schema's are derivatives of JSON-LD. The link you shared probably refers to one such derivative.
